### PR TITLE
Creates a script to normalize the values of users.gender.

### PR DESCRIPTION
**Pull request by ashercodeorg**
The script is expected to touch approximately O(730K) users and take strictly less than eight hours (based on #8383 taking five hours) if `start_index = 0` and `end_index = infinity`.

Per the comment in the script, another PR will establish the corresponding migration (to be run after this script runs) so as to assure the same invariant in all environments.

LOCAL TESTING (before and after):

```
irb(main):030:0> User.all.pluck(:id, :name, :gender)
   (0.5ms)  SELECT `users`.`id`, `users`.`name`, `users`.`gender` FROM `users` WHERE `users`.`deleted_at` IS NULL
=> [[1, "AMK Local Admin", nil], [2, "gender m", "m"], [3, "gender f", "f"], [4, "gender d", "d"], [5, "gender -", "-"], [6, "gender space", " "], [7, "gender q", "q"]]
irb(main):031:0> User.all.pluck(:id, :name, :gender)
   (1.2ms)  SELECT `users`.`id`, `users`.`name`, `users`.`gender` FROM `users` WHERE `users`.`deleted_at` IS NULL
=> [[1, "AMK Local Admin", nil], [2, "gender m", "m"], [3, "gender f", "f"], [4, "gender d", nil], [5, "gender -", "-"], [6, "gender space", nil], [7, "gender q", nil]]
```

SCOPE:

```
SELECT COUNT(0)
FROM users
WHERE gender IS NOT NULL AND gender != '-' AND gender != 'f' AND gender != 'm';
+----------+
| COUNT(0) |
+----------+
|   729461 |
+----------+
```

### CODE REVIEW

File: bin/oneoff/normalize_gender [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+# of its execution in production, followed by a migration performing the same
+# task so as to assure good data in all environments.
+
+require_relative '../../dashboard/config/environment'
+require src_dir 'database'
+
+User.with_deleted.
+  where('gender IS NOT NULL').
+  where('gender != "-" AND gender != "f" AND gender != "m"').
+  update_all(gender: nil)
```
Comment (ashercodeorg) [inline bin/oneoff/normalize_gender:28]: Per documentation (http://apidock.com/rails/v4.0.2/ActiveRecord/Relation/update_all), note that this does not trigger callbacks or validations.

**Comment by ashercodeorg** [comment]
CC @laurelfan

**Comment by laurelfan** [comment]
do you have an estimate for how long this will take to run in prod?

**Comment by ashercodeorg** [comment]
Updated the PR comment to include a rough estimate and a statement of scope.

**Comment by laurelfan** [comment]
hm, 8 hours... what do you think the impacts of having such a long running single statement are? (like, will it lock, will it fail because rows were changed...), would it be better to do it in batches?

**Comment by Hamms** [comment]
I seriously doubt this will actually take on the order of hours; the referenced script was creating O(number of users) MySQL UPDATEs, I believe this one will create only one.

With that belief in place, this LGTM!

**Comment by laurelfan** [comment]
do we have any basis for that besides "mysql would be really dumb if it did"? Mysql is just dumb sometimes (just asking because we were surprised that the last one took 5h).

If we need to get more data points to narrow the bounds better than <1-8 hours it would be easy (but a bit inconvenient) to create a test db from a backup snapshot and and run the generated query on it. (and while you are at it, run the update again after it finishes to verify that it's fast if it needs to update 0 rows -- I am way less skeptical about this but if that is right it should take no time to hit up arrow and return). Your call based on how sure this will not cause any problems ;)

**Comment by ashercodeorg** [comment]
That this operates in a single MySQL statement (per the [documentation](http://apidock.com/rails/v4.0.2/ActiveRecord/Relation/update_all))  is both a blessing (should run faster) and a curse (everything is in a single statement, more likely to be a problem) in my opinion. It is easy enough to add a `start_index` and `end_index` as suggested, so I did.

**Comment by ashercodeorg** [comment]
Given that (a) we can run the script off hours, (b) the script is idempotent, (c) a failed `update_all` shouldn't be harmful (right?), and (d) we have a "strict" upper bound on how long it should take, I don't see there being any harm in running the script a million (e.g., `end_index - start_index`) users at a time (verifying the users to update are uniformly distributed).

**Comment by Hamms** [comment]
Note that the last one is not operating purely at the MySQL level; it drops down into the ruby layer and calls `save` manually for each user, each of which will make its own MySQL request. Totally reasonable that that one would take ages.

That being said, batching is always good.

**Comment by ashercodeorg** [comment]
Unless there is objection, I'll submit as-is (with batching), running it in "(large) bite-size" batches on production.

**Comment by ashercodeorg** [comment]
Running with `end_index - start_index = 1M` was successful, finishing in O(30 seconds).

Running with `end_index - start_index = 10M` was _NOT_ successful, resulting in stealing the `users` table lock long enough to have production user queries timeout.

**Pull request merged**